community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/NodeRelationshipCache.java

```diff
@@ -117,28 +193,89 @@ public long getAndPutRelationship( long nodeId, int type, Direction direction, l
 
         ByteArray array = this.array.at( nodeId );
         long existingId = all48Bits( array, nodeId, SPARSE_ID_OFFSET );
-        if ( isDense( array, nodeId ) )
+        boolean dense = isDense( array, nodeId );
+        boolean wasChanged = markAsChanged( array, nodeId, changeMask( dense ) );
+        markChunkAsChanged( nodeId, dense );
+        if ( dense )
         {
             if ( existingId == EMPTY )
             {
-                existingId = relGroupCache.allocate( type, direction, firstRelId, incrementCount );
+                existingId = relGroupCache.allocate();
                 setRelationshipId( array, nodeId, existingId );
-                return EMPTY;
+                wasChanged = false; // no need to clear when we just allocated it
             }
-            return relGroupCache.putRelationship( existingId, type, direction, firstRelId, incrementCount );
+            return relGroupCache.putRelationship( existingId, direction, firstRelId, incrementCount, wasChanged );
         }
 
         // Don't increment count for sparse node since that has already been done in a previous pass
         setRelationshipId( array, nodeId, firstRelId );
-        return existingId;
+        return wasChanged ? EMPTY : existingId;
+    }
+
+    private void markChunkAsChanged( long nodeId, boolean dense )
+    {
+        byte mask = chunkChangeMask( dense );
+        if ( !chunkHasChange( nodeId, mask ) )
+        {
+            int chunk = chunkOf( nodeId );
+            if ( (chunkChangedArray[chunk] & mask) == 0 )
+            {
+                // Multiple threads may update this chunk array, synchronized performance-wise is fine on change since
+                // it'll only happen at most a couple of times for each chunk (1M).
+                synchronized ( chunkChangedArray )
+                {
+                    chunkChangedArray[chunk] |= mask;
+                }
+            }
+        }
+    }
+
+    private int chunkOf( long nodeId )
+    {
+        return toIntExact( nodeId / chunkSize );
+    }
+
+    private static byte chunkChangeMask( boolean dense )
+    {
+        return (byte) (1 << (dense ? 1 : 0));
+    }
+
+    private boolean markAsChanged( ByteArray array, long nodeId, long mask )
+    {
+        long bits = array.getInt( nodeId, SPARSE_COUNT_OFFSET ) & 0xFFFFFFFF;
```

Comment:
This is wrong. You need to AND with `0xFFFFFFFFL` instead of `0xFFFFFFFF`.

Proof:

```
int x = 0xF0_00_00_00;
long X = x & 0xFFFFFFFF;
long Y = x & 0xFFFFFFFFL;
System.out.println( "X = " + X );
System.out.println( "Y = " + Y );
```

Prints

```
X = -268435456
Y = 4026531840
```

Intellij warns about this. Maybe you can make Eclipse do the same.